### PR TITLE
Admin merchants ar

### PR DESCRIPTION
**Pull request by Scott-Borecki**
### What I Did:
- Completed the following user stories:
  - #18 
  - #19 
  - #20 
  - #21 
  - #22

### Road Blocks
- ActiveRecord

### Test Coverage
- [X] I have tested this code with RSpec
  - Features: 213/214 (99.53%)
  - Models: 120/122 (98.36%)
- We can check out what methods are not being tested.

### Questions/Notes:
- None

### CODE REVIEW

File: app/models/customer.rb [hunk 1]
```diff
@@ -7,4 +7,13 @@ class Customer < ApplicationRecord
 
   validates :first_name, presence: true
   validates :last_name, presence: true
+
+  def self.top_five_by_successful_transactions
+    joins(:transactions)
+      .select('customers.*, count(transactions.id) AS count')
+      .where(transactions: { result: :success})
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply app/models/customer.rb:14]: Updated

File: spec/models/customer_spec.rb [hunk 2]
```diff
@@ -1,6 +1,27 @@
...
+  describe 'object creation for tests' do
+    specify { expect(Customer.all.count).to be_positive }
+    specify { expect(Merchant.all.count).to be_positive }
+    specify { expect(Item.all.count).to be_positive }
+    specify { expect(Invoice.all.count).to be_positive }
+    specify { expect(Transaction.all.count).to be_positive }
+    specify { expect(InvoiceItem.all.count).to be_positive }
+
+    it 'can build a customer' do
+      customer_1 = create(:customer)
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/models/customer_spec.rb:18]: Update

File: spec/create_test_objects.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    i = 0
+    j = 0
+    Invoice.all.each do |invoice|
+      k = 0
+      3.times do
+        invoice.invoice_items.create!(
+          quantity: invoice_item_data[j][0],
+          unit_price: (i + k >= 19) ? items[i - k].unit_price : items[i + k].unit_price,
+          status: invoice_item_data[j][2],
+          item_id:( i + k >= 19) ? items[i - k].id : items[i + k].id
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/create_test_objects.rb:76]: Updated

File: app/controllers/admin_controller.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class AdminController < ApplicationController
+  def index
+    @top_customers = Customer.top_five_by_successful_transactions
```
Comment (kabcouwer) [inline app/controllers/admin_controller.rb:3]: Nice use of this method in the controller - Caroline and I could implement into our too

File: app/views/admin/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      <% @incomplete_invoices.each do |incomplete_invoice| %>
+        <p>
+          Invoice #<%= link_to incomplete_invoice.id, admin_invoice_path(incomplete_invoice) %>
+          <%= incomplete_invoice.formatted_time %>
+        </p>
+      <% end %>
+    </div>
+    <div id='top-five-customers' class="col-lg-4 col-sm-6">
+      <h3>Top Customers</h3>
+      <ol>
```
Comment (kabcouwer) [inline app/views/admin/index.html.erb:25]: Love the ordered list implementation

File: app/views/admin/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        <% @top_customers.each do |customer| %>
+          <li>
+            <section id='top-customer-<%= customer.id %>'>
+              <%= customer.first_name %> <%= customer.last_name %> - <%= customer.count %> purchases
```
Comment (kabcouwer) [inline app/views/admin/index.html.erb:29]: Could create a full name method for customer.. if we have time lol

File: spec/features/admin/index_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      end
+
+      describe 'when I look in the top 5 customers section' do
+        it 'displays the names and number of puchases of the top 5 customers' do
+          top_customers = Customer.top_five_by_successful_transactions
+
+          within '#top-five-customers' do
+            expect(page).to have_content('Top Customers')
+            top_customers.each do |customer|
+              expect(page).to have_content("#{customer.first_name} #{customer.last_name} - #{customer.count} purchases")
```
Comment (kabcouwer) [inline spec/features/admin/index_spec.rb:50]: very cool!

**Review by kabcouwer — APPROVED** [review]
Testing looks really great! Super thorough. I'm going to try out doing loops in my testing like y'all did because its so awesome.

Not sure whats going on with styling but looks dope!

Nice work! 2 very enthusiastic thumbs UP

**Pull request merged**